### PR TITLE
memory headroom canonical policy support enable buffer

### PR DESCRIPTION
**Pull request by luomingmeng**
#### What type of PR is this?
Features
#### What this PR does / why we need it:
Current canonical memory headroom policy only consider the memory estimation of non-reclaimed QoS container, and uses the RSS memory to estimate, which may trigger memory reclamation frequently when shared or dedicated QoS container use much more cache and shmem in high load case, so we also need consider them in memory estimation and add extra memory margin.
In the low load case, the machine has enough free memory, the extra memory margin or cache can also oversold to reclaimed QoS worklaod to increase memory utilization.

### CODE REVIEW

File: cmd/katalyst-agent/app/options/sysadvisor/qosaware/resource/memory/headroom/policy_canonical.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	defaultStaticMemoryOversold            = 20 << 30 // 20GB
+
+	defaultCacheOversoldRate                     = 0
+	defaultCacheOversoldMemoryCPURatioUpperBound = 6.0
+	defaultCacheOversoldMemoryCPURatioLowerBound = 3.5
+)
+
+type MemoryPolicyCanonicalOptions struct {
+	EnableBuffer                          bool
+	MemoryFreeEstimationUtilization       float64
```
Comment (waynepeking348) [inline cmd/katalyst-agent/app/options/sysadvisor/qosaware/resource/memory/headroom/policy_canonical.go:37]: // add comments here to explain we will choose the max of them
type MemoryUtilBasedOptions struct {
   FreeBasedRatio
   CacheBasedRatio
   StaticBasedCapacity
   
   Enabled
   CPUMemRatioLowerBound
   CPUMemRatioUpperBound
}

File: pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go [hunk 2]
```diff
@@ -29,71 +30,263 @@ import (
...
+}
+
+type regionMemoryMetrics struct {
+	cache float64
+	shmem float64
+	rss   float64
+}
+
+// getRegionMemoryMetrics get region memory metrics from meta server
+func (p *PolicyCanonical) getRegionMemoryMetrics(filter func(pod *v1.Pod) bool) (*regionMemoryMetrics, error) {
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go:251]: getRegionMemoryMetrics -> getMemoryMetrics

File: pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go [hunk 3]
```diff
@@ -29,71 +30,263 @@ import (
...
+	cpuCapacity := p.metaServer.MachineInfo.NumCores
+	memoryCapacity := p.metaServer.MemoryCapacity
+	if cpuCapacity == 0 {
+		return 0, fmt.Errorf("cpu capacity is 0")
+	}
+
+	return float64(memoryCapacity) / 1024 / 1024 / 1024 / float64(cpuCapacity), nil
+}
+
+type regionMemoryMetrics struct {
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go:244]: regionMemoryMetrics -> memoryMetrics

File: pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go [hunk 4]
```diff
@@ -29,71 +30,263 @@ import (
...
 
 type PolicyCanonical struct {
 	*PolicyBase
 
 	// memoryHeadroom is valid to be used iff updateStatus successes
-	memoryHeadroom float64
-	updateStatus   types.PolicyUpdateStatus
+	memoryHeadroom               float64
+	updateStatus                 types.PolicyUpdateStatus
+	qosConfig                    *generic.QoSConfiguration
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go:48]: divide into groups

File: pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go [hunk 5]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package headroompolicy
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_canonical.go:18]: move to another file

**Review by waynepeking348 — APPROVED** [review]

**Review by sun-yuliang — APPROVED** [review]

**Pull request merged**